Add missing dependencies of magento/framework

GordonLesti opened the pull request:
If I require `magento/framework` in my module, cause I'm using class `Magento\Framework\HTTP\PhpEnvironment\Request` for example. My unit tests are breaking cause `Zend\Http\Header\HeaderInterface` is used in `Magento\Framework\HTTP\PhpEnvironment\Request`, but the class `Zend\Http\Header\HeaderInterface` can not be found. This forces me to require `zendframework/zend-http` in my module without using any class from this library.

GordonLesti committed:
magento/framework should require zendframework/zend-http

ihor-sviziev commented on lib/internal/Magento/Framework/composer.json:
```diff
@@ -20,7 +20,8 @@
         "ext-xsl": "*",
-        "symfony/process": "~2.1"
+        "symfony/process": "~2.1",
+        "zendframework/zend-http": "~2.4.6"
```
I think no need to prevent updating to 2.5 and more. So better to use `~2.4`. If it should be more than 2.4.6 - `~2.4 >=2.4.6`

joshdifabio replied:
`^2.4.6` can be used as shorthand for `>=2.4.6,<3`.

GordonLesti replied:
I have taken the requirement from `magento/magento2ce` [composer.json](https://github.com/magento/magento2/blob/2.0.4/composer.json#L39) and that is `"zendframework/zend-http": "~2.4.6"` at the moment.

GordonLesti commented:
I wasn't completly sure. That's why I haven taken the required version of the composer.json from `magento/magento2ce`.

vkublytskyi commented:
@GordonLesti, thank you for your contribution.

Unfortunately there are another missed dependencies of Magento Framwork that are declared only in root `composer.json` file. We've created internal ticket `MAGETWO-51702` to review and fix dependencies of Magneto Framework.

But if you are interested and may help us with resolving this issue we will be very thankful.

GordonLesti commented:
Yes, my unit tests just forced me to require `zendframework/zend-http` without using any class from that library so I have only added `zendframework/zend-http` at the moment. I check also for other dependencies & will update my contribution as sson as possible.

GordonLesti committed:
add all dependencies of magento/framework

GordonLesti commented:
I have now added all not dev dependencies of `magento/framework` to `lib/internal/Magento/Framework/composer.json`. For every requirement I have added, I have an example usage in this list.
- `Zend\Code\Generator\MethodGenerator` in `Magento\Framework\Code\Generator\ClassGenerator`
- `Zend\Stdlib\Parameters` in `Magento\Framework\App\Test\Unit\Request\HttpTest`
- `Zend\Mvc\Application` in `Magento\Framework\Console\Cli`
- `Zend\Crypt\Utils` in `Magento\Framework\Encryption\Helper\Security`
- `Zend\Escaper\Escaper` in `Magento\Framework\Escaper`
- `Zend\Stdlib\Glob` in `Magento\Framework\Filesystem\Glob`
- `Zend\Http\Header\HeaderInterface` in `Magento\Framework\HTTP\PhpEnvironment\Request`
- `Zend\Uri\UriFactory` in `Magento\Framework\HTTP\PhpEnvironment\Request`
- `Zend\Validator\Hostname` in `Magento\Framework\Session\Config\Validator\CookieDomainValidator`
- `Psr\Log\LoggerInterface` in `Magento\Framework\Phrase\Renderer\Inline`
- `Cm\RedisSession\Handler\ConfigInterface` in `Magento\Framework\Session\SaveHandler\Redis`
- `Composer\Autoload\ClassLoader` in `Magento\Framework\Autoload\ClassLoaderWrapper`
- `Monolog\Formatter\LineFormatter` in `Magento\Framework\Logger\Handler\Base`
- `Less_Parser` in `Magento\Framework\Css\PreProcessor\Adapter\Less\Processor`
- `CSSmin` in `Magento\Framework\Code\Minifier\Adapter\Css\CSSmin`
- `Symfony\Component\Console\Input\ArrayInput` in `Magento\Framework\Composer\DependencyChecker`
- `JShrink\Minifier` in `Magento\Framework\Code\Minifier\Adapter\Js\JShrink`
- `Magento\Composer\InfoCommand` in `Magento\Framework\Composer\MagentoComposerApplicationFactory`

vkublytskyi commented:
@GordonLesti Thank you for so quick response. We will process your PR

mazhalai commented:
@GordonLesti please sync with the develop branch and rerun travis builds.

GordonLesti committed:
Merge remote-tracking branch 'upstream/develop' into require_zend_http

GordonLesti commented:
@mazhalai `develop` is merged & travis builds

vrann commented:
@GordonLesti sorry for asking to do it again, but can you please merge with the latest develop?

GordonLesti commented:
@vrann no problem, will merge develop tomorrow and ping you again.

GordonLesti committed:
Merge remote-tracking branch 'upstream/develop' into require_zend_http

Conflicts:
	lib/internal/Magento/Framework/composer.json

GordonLesti committed:
alphabetical order of third party dependencies

GordonLesti committed:
update version of dependencies

GordonLesti commented:
Hello @vrann 
What have I done since last merge of `develop` branch.
* Merged `develop` again
* Ordered third party dependencies in alphabetical order
* Checked again if all example of code dependencies from my comment above are persist
* Updated versions of dependencies with the versions from [composer.json](https://github.com/magento/magento2/blob/be7cbf26895beae6847d4445e1f37b4714cbf5cd/composer.json) of `magento/magento2ce`

What I have not done, or what may should be done in future
* Check for new code dependencies that aren't mentioned yet in `composer.json` of `magento/framework`
